### ISSUE TITLE
Native support for Windows 10 on ARM (ARM64 Win32)

### ISSUE THREAD

**Issue by jkunkee** [notepad-plus-plus/notepad-plus-plus#5158]
While I'm filing this feature request as a long-time fan of Notepad++, please note that I'm also a Microsoft software engineer whose day job is expanding the Windows 10 on ARM (ARM64, both Win32 and UWP) software ecosystem.

I use the x86 ("32-bit") version of Notepad++ on a regular basis on a Windows 10 on ARM device, and it works very well. There is also a wide variety of compiled plugins available for it. There's a solid argument for not doing the extra work to support another architecture, though I see x64 hasn't been out of the question. :)

Notepad++ is a wonderfully light application, so the benefits of native ARM64 support are simple:

 * Lower power consumption
    * The idle loop takes ~0.2% CPU, but I tend to leave Notepad++ running. A lot.
 * Slightly better responsiveness
    * Seriously, when I started it a few minutes ago typing was just slightly slow to render. I tried again and couldn't notice the difference again. It's slight.

With Visual Studio 2017 15.9+ and Windows SDK 10.0.17763.0, ARM64 Win32 is fully supported (no `WindowsSDKDesktopARMSupport` even). Projects often just need to add an ARM64 configuration to build, the biggest questions being about binary dependencies and assembly. 

I may eventually get to prepping a PR, but in the meantime I'm happy to answer questions about what this requires.

**Comment by zufuliu** [comment]
Hi @jkunkee 

Seems arm and arm64 SDK is different from x86 and x64, files count from 
```
Windows Kits\10\Lib\10.0.17763.0\um
    arm              348
    arm64            358
    x64              474
    x86              475
```
and shlwapi.lib is missing from arm and arm64 folder.

**Comment by zufuliu** [comment]
After repair Windows SDK, all lib is here:
```
Windows Kits\10\Lib\10.0.17763.0\um
    arm              444
    arm64            454
    x64              474
    x86              475
```
But "Inherited values" for linker input is different:
for x64:
```
gdi32.lib
winspool.lib
comdlg32.lib
advapi32.lib
shell32.lib
ole32.lib
oleaut32.lib
uuid.lib
odbc32.lib
odbccp32.lib
```

and for arm64:
```
kernel32.lib
user32.lib
```

**Comment by jkunkee** [comment]
Interesting. The 'inherited values' may not be intended to be similar; I'll ask around. The lib and Redist folders won't be completely equivalent, though: Windows 10 on ARM first shipped with Windows 10 1709, so DLLs that ship with Windows 10 and not Windows 8 aren't needed and LIBs for certain old features are omitted.

Outside the short list of exceptions, though, if it's a LIB in the x64 SDK it should be in the ARM64 SDK. Feedback Hub is the best way I'm aware of for reporting discrepancies and changing that short list would probably require a UserVoice issue to get attention, but as I said I'll ask around to see if the two you noted here are expected. A 20 LIB differential is bigger than I expected.

**Comment by jkunkee** [comment]
After some digging, the 20 or so libs are unavailable features on current ARM64 platforms (OpenGL, Hyper-V) or purposely dropped libs. I'm glad the install repair fixed that up for you.

The Linker Input inherited value discrepancy is a known bug that is being fixed. In `<vsinstalldir>\Common7\IDE\VC\VCTargets\Platforms\ARM64\PlatformToolsets\v141\Toolset.props`, you'll find that `MinimalCoreWin` is being incorrectly set to 'true'.

**Comment by jkunkee** [comment]
Windows 10 1709 was the first Windows desktop OS on ARM64, so yes, 0x0A00 is the right value for it.

I can give Notepad2 a shot, yes, but Notepad++ is my daily driver. (Doesn't Notepad2 ship with Git for Windows as the default editor?)

**Comment by jkunkee** [comment]
I have a coworker that didn't have much trouble with Boost when he did an ARM64 build of Notepad++ some time back. I'll ask around.

**Comment by MIvanchev** [comment]
Hi everybody, just providing hopefully usable info to the discussion. I maintain an [ARM32 build of Notepad++](https://github.com/MIvanchev/notepad-plus-plus/tree/boost_appveyor) featuring an extended AppVeyor build which builds Boost.Regex and Scintilla automatically. The changes could be used for ARM64 as well. The plugins will indeed be a major bottle neck, but if we combine our effort it's doable =). Cheers!

**Comment by jkunkee** [comment]
@zufuliu - thanks for the history lesson!

It looks like there are already scripts for building Notepad2 for ARM64 Windows. Is it a question of automating it so it's part of official releases? MIvanchev seems to have gotten AppVeyor to produce ARM32 builds, so his experience might be helpful. If it's something else, open an issue and tag me in a comment; I'm happy to take a look.

@qz3fwd - Wow! Awesome! Do let us know how it worked out. :)

Your pain around adding LIBs to the linker config is a known bug in the SDK (MinimalCoreWin). It is rather annoying for now but should be fixed with the next SDK version.

I understand wanting to avoid installer work at the beginning. When it does come up, note that (at least) tip-of-tree NSIS has a function named `IsNativeARM64` that lets an x86 installer detect when it's running on ARM64. (My probably-too-thorough version of this trickery [involved some arcane NSIS runework](https://github.com/OpenVPN/tap-windows6/pull/57/files).)

@MIvanchev - Have you asked the greater Notepad++ community if your changes and architecture addition could be upstreamed to the primary project? I'm hoping that at least ARM64 builds will be available on the main site sooner or later, complete with update notifications.

Also, I see that your AppVeyor script still uses the VS2015 image. For ARM64 you should need the VS2017 image since that is when the non-beta ARM64 tooling landed.

About plugins: I'm not at all familiar with N++ plugin development. Is there a sample repo the other plugins are based off of? Adding ARM64 support there might help plugin maintainers make the leap. How is the plugin port problem being handled for x64?

**Comment by zufuliu** [comment]
@jkunkee 
ARM64 build for Notepad2 is available at https://ci.appveyor.com/project/zufuliu/notepad2.

**Comment by jkunkee** [comment]
@zufuliu - That was what I was missing! I just stitched it into my local workflow. The Release version can open, highlight, and edit CMD, PS1, and HTML files on my HP Envy x2. (I'll definitely revisit this thread when there's a good low-cost Windows 10 on ARM(64) automated testing story around.)

**Comment by qz3fwd** [comment]
>  building Notepad2 for ARM64 Windows. Is it a question

I have Notepad++ 7.5.7 built as a native ARM64 exe and running fine on my HP Envy X2 snapdragon 835 CPU tablet. C++ source files look beautiful on it and it takes up like 0.1% CPU load. A couple points.
I had to build libcurl.dll from source and there is no ARM support in the curl makefile. Easy fix though.
Boost version referenced by Notepad++ source is "old" and has no knowledge of the VS compiler version. Again another easy fix. I built a static lib solution for the libregex library instead of wrestling with building ALL of boost for ARM64 which is not straightforward, nor well supported. I have always though the boos libraries are not straightforward to build over the last 10+ years regardless of cpu or VS version-they have their own custom build system with several different versions over the years-b2 or jam.... I had to disable the SciLexer.dll signing verification in the sources (as there is no signed ARM64 compiled DLL I am aware of) otherwise the program will not function.  

I still need to look through the source code and disable startup update check. I got an error message gup.exe failed...... (I didn't build gup.exe but it would have failed anyways since the updated cannot find a newer ARM64 build). I also need to look into building plugins native to ARM64. Maybe tonight.

Anyone interested in the binaries? I can upload them if any interest-not sure where.

Long live Windows on ARM.

In summary I am LOVING having Notepad++ on my ARM64 tablet. No emulation for me of x86 apps-kinda defeats the entire purpose of the device and kills battery like and performance......

**Comment by qz3fwd** [comment]
Vcpkg failed miserably building both boost regex as well as angle-sadly. Even though I used the arm64 windows triplet the logs showed it apparently still trying to build an x86 version for boost regex. Guess the ARM64 support is very rudimentary.

**Comment by chcg** [comment]
@qz3fwd @jkunkee See e.g. https://github.com/npp-plugins/plugintemplate as starting point for plugin development used by various plugins. There are also plugins not written in C++, but C#, Delphi, etc.

See also https://github.com/chcg/nppPluginList/tree/appveyor_for_arm

**Comment by sredna** [comment]
@jkunkee Regarding the idle loop CPU usage, do you happen to have the DSpellCheck plug-in? It seems to fire a WM_TIMER even when N++ is not active.

**Comment by MetaChuh** [comment]
we will currently not pursue arm development at the main project.
please feel free to continue this dev discussion here, or at our notepad++ community.

thank you for your help and best regards.

**Comment by jkunkee** [comment]
@sredna, I have a two-part answer: I do have DSpellCheck, but I believe it's the newer version without the WM_TIMER issue--that said, the idle CPU usage has dropped.

@MetaChuh, thanks for the official response. One of these days I'll get to looking at what @chcg posted, if someone else doesn't first...

**Comment by NiQ1** [comment]
Hi,
Please reconsider this. I've managed to build the entire thing including Boost regex within a couple of hours. The only necessary changes are disabling a couple of warnings (e.g. GetVersionEx deprecated etc.), changing a few #ifndefs in the boost headers and as long as no official builds are available then also removing the signature check.

Binaries of Notepad++ ARM64 as well as some other ARM64 ports are available here: https://mega.nz/#F!PclhDIRB!-yhBZ6UM7S596ijNU3dx0A

(if you want the source code changes contact me but they are so obvious that you probably won't even need them).

I know plugins will be a problem but we have to begin somewhere.

**Comment by FilipeRamalho** [comment]
I wanted to confirm if arm64 development won't be pursued after the announcement of the Surface Pro X.

**Comment by hernancmartinez** [comment]
@jkunkee Seconded by the comment above (I just pre ordered a Surface Pro X btw). Did you see that Boost now has proper support for ARM64 ? https://www.boost.org/doc/libs/1_71_0/doc/html/intrusive/release_notes.html#intrusive.release_notes.release_notes_boost_1_65_00 I'll check how easy is to build Boost.Regex later 

Was Scintilla easy to build?

**Comment by donho** [comment]
@jkunkee 
If Windows 10 ARM means UWP system ?

**Comment by zufuliu** [comment]
@donho I think it's not about UWP. It's just compiling Notepad++ and depended libraries and plugins with ARM64 compiler:
* Add ARM64 platform (e.g.: copy from x64)
* Change `_WIN32_WINNT`, `WINVER` or `NTDDI_VERSION` to match Win10 (0x0A00)
* Add `<MinimumRequiredVersion>10.00</MinimumRequiredVersion>` for linker.

See people runs Notepad2 on ARM64
![Notepad2 AMR64](http://www.technique.cz/storage/Notepad2_10.jpg)

**Comment by donho** [comment]
@zufuliu Thank you for your answer. So from the previous post of this issue, Notepad++ x64 binary can work under ARM64 device but with some performance issues, I do understand the need. The question that I ask myself is, how many windows run on ARM device? Is the effort (for boost/scintilla/gup/plugins) worth ?

**Comment by jkunkee** [comment]
(in case my answer helps) No, ARM64 Windows does not mean UWP, but you can build ARM64 UWP.

Subtlety: only the 32-bit x86 version runs on Windows 10 on ARM(64), not x64. If Microosft adds and ships x64 support, from what I know of emulation it won't be fast.

I'm not allowed to share numbers, sadly. My impression is that it's doing quite well even before the Surface Pro X ships, but it won't be hitting one billion devices on its own?

Anecdotally (!), all of the ARM64 devices I get to play with end up with Notepad++ installed on them. I know a few devs that love their ARM64 devices and use Notepad++. There's some interest on this thread. Yay anecdata! :)

My impression so far has been (please correct me if I'm wrong!):

- Scintilla doesn't need work
- Boost needs a version bump and a config change
- plugins are the most challenging

In-box plugins are easy, but plugins published by others are hard simply because they will need to make config changes, turn on building for ARM64, then publish ARM64 binaries. If they have dependencies, those may need porting too.

**Comment by NiQ1** [comment]
@donho There's no relationship between ARM64 and UWP, this is not Windows RT all over again. It is full Windows 10 and you can't even tell the differences just by looking at the desktop. The entire Win32 API is available (very few exceptions, like OpenGL but nothing that's relevant to us).
In a previous message I've posted a link to an ARM64 build I've compiled myself with hardly any code changes (the very few modifications were related to the switch in VC version rather than the actual platform). Again I will mention that it is definitely not UWP - you won't be able to tell that you're on ARM64 just by opening and working with Notepad++.
The only downside is plugins - ARM64 binaries can't load x86 DLLs (and vice versa), just the same as x64, which means that an ARM64 build of Notepad++ could only load plugins that were also built as ARM64. It'll be up for people to decide if they want to use it or run the x86 version under emulation (and suffer the performance penalty).
I am in favor of an ARM64 build, not because it's so much needed that we have to put a lot of effort, but rather because there's almost zero effort in porting it (as I said I was able to build it within a couple of hours).

x64 cannot run on ARM64 at all, this is not supported an Microsoft has confirmed it probably never will be, due to both performance and complexity of implementation. x86 works but since it's full CPU emulation it gets a significant performance hit.

**Comment by qz3fwd** [comment]
I agree and support a native ARM64 win32 binary build of Notepad++. I myself built Notepad++ about a year ago for my HP Envy X2 and current Lenovo C630. The build was almost effortless. The hardest part was disabling signing checks for scintilla.

Actually the hardest part was extracting boost thread and building a dll myself, but now boost supports building a Windows ARM64 target out of the box.

Not sure how many Win10 ARM64 users there are but after the Microsoft device starts shipping the numbers will be much larger.  ARM64 Win10 supports basically the entire Win32 API.

Thanks for reconsidering.

Yes lack of opengl has caused me problems building the 2 main commercial MFC toolkits, but that is beside the point here.

If we could add a ARM64 build target that would be great.

**Comment by donho** [comment]
@jkunkee @NiQ1 @qz3fwd 
Thank you for your information.
Plugins part could be challenging but I'm not worry about it. They will catch up if there's the need from users - surely there will be the online tutorial to help them do the port from Notepad++ community side.

OTOH, NSIS installer part could be an impact for ARM distribution. I just google it and didn't see the ARM possibility for NSIS, and I'm not ready to switch to MSI.

What I need now is rather the instructions (step by step) to compile PCRE of boost (with a specific version), and Scinitilla manually for a POC. Could you provide the instructions?

**Comment by qz3fwd** [comment]
As I recall from last year, I used BCP to isolate the boost dependency and built the lib as a VS dll project. Now boost build natively builds ARM64 libs so this step could be simplified.

I will check tonight after work what adjustments to the build process I made for scintilla component and try to post back tonight.

**Comment by sredna** [comment]
@donho

Most of NSIS compiles for ARM64 but I don't know if it actually works because I don't have hardware to test on.

As it stands now, you will have to use a x86 32-bit installer but that should not be a problem. NSIS does have detection for ARM64 in the x86 version in x64.nsh: 

```NSIS
Function .onInit
${IfNot} ${IsNativeARM64}
  MessageBox MB_ICONSTOP "ARM64 version blah blah!"
  Quit
${EndIf}
FunctionEnd
```

**Comment by robmikh** [comment]
I was able to get this building by using the current 1.71.0 version, using the same directions in this project's README. The differences were:

- I ran everything in the "x64_arm64 Cross Tools Command Prompt for VS 2019"
- When compiling boost, I added `architecture=arm` (not sure if it did anything)
- When opening notepadPlus.vcxproj I had to update to platform tools v142 (I only have v142 installed, not sure if necessary)
- Added an ARM64 platform under the Configuration Manager in Visual Studio

I was able to build the "Unicode Debug" configuration for ARM64. The release variant didn't want to work for me, it wasn't able to match SciLexter.dll's signature. I'm not sure how much of this is due to changing the project to use v142 and the latest sdk, and how much of it is from targeting ARM64.

In the end I have a debug ARM64 build of Notepad++ that I'm able to run on my Surface Pro X, so I'm happy. Just thought I'd add my experience in case anyone else came across this thread like I did.

**Comment by chcg** [comment]
@robmikh See https://github.com/notepad-plus-plus/notepad-plus-plus/pull/6196 for the VS project changes + appveyor CI build. For the release version the dll needs to be signed by donho with his private key. I think this is no issue of v142 or ARM64 as this is also happening for x86 and x64 release versions.

**Comment by minecraftchest1** [comment]
When I saw this thread, I was thinking "just build it using c++/clr and run it as a .net assembly." I'll try it when I can. I will try building it for raspberrian on my raspberry pi 3b+. Then we can have releases for windows, linux, and android.
Also, why don't we just build our own installer? I'll try that as well.

**Comment by jeffthompsonsd** [comment]
Any progress on this? Would love a native arm64 NP++

**Comment by jeffthompsonsd** [comment]
I was able to get a debug arm64 version built like @robmikh but none of the dlls that would end up in the plugins directory were built, how are those generated? mimeTools, nppConverter, nppExport

but at least ive got native np++ on my spx!

**Comment by chcg** [comment]
@jeffthompsonsd These are plugins from different repos, see e.g. https://github.com/npp-plugins/mimetools or https://github.com/chcg/NPP_ExportPlugin (with https://github.com/chcg/NPP_ExportPlugin/releases/download/0.3.0/NppExport_0.3.0_arm64.zip) or https://github.com/notepad-plus-plus/nppPluginList with some testbranch for arm64 at https://github.com/chcg/nppPluginList/tree/appveyor_for_arm)

**Comment by jeffthompsonsd** [comment]
> @jeffthompsonsd These are plugins from different repos, see e.g. https://github.com/npp-plugins/mimetools or https://github.com/chcg/NPP_ExportPlugin (with https://github.com/chcg/NPP_ExportPlugin/releases/download/0.3.0/NppExport_0.3.0_arm64.zip) or https://github.com/notepad-plus-plus/nppPluginList with some testbranch for arm64 at https://github.com/chcg/nppPluginList/tree/appveyor_for_arm)

Gotcha thanks!

**Comment by donho** [comment]
Could you guys try the following Notepad++ ARM64 and confirm us if it work under your ARM device please?
[nppArm64.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6240862/nppArm64.zip)

The result comes from the PR of @chcg :
https://github.com/notepad-plus-plus/notepad-plus-plus/pull/9594

If it works, we'll have soon Notepad++ ARM 64 bits release.

**Comment by robmikh** [comment]
I gave it a go on a Surface Pro X and it launched fine. I also gave the C syntax highlighting a quick look, seemed everything was working.

Thanks for the working on this!

**Comment by savelov** [comment]
> 
> 
> Could you guys try the following Notepad++ ARM64 and confirm us if it work under your ARM device please?
> [nppArm64.zip](https://github.com/notepad-plus-plus/notepad-plus-plus/files/6240862/nppArm64.zip)
> 
> The result comes from the PR of @chcg :
> #9594
> 
> If it works, we'll have soon Notepad++ ARM 64 bits release.

works for me on ASUS TP370QL

**Comment by hmartinez82** [comment]
Works for me on Surface Pro X

**Comment by ntw-au** [comment]
Thank you! 😊 Works for me on Samsung Galaxy Book2, having given a quick test for open, save, rename, display font, syntax highlighting and changing encoding and line endings.

**Comment by atl068** [comment]
Works well on my Galaxy Book 2 too. Thank you!

**Comment by ch3mn3y** [comment]
Works amazing on Lumia 950XL. Many thanks

**Comment by donho** [comment]
@ch3mn3y 
> Works amazing on Lumia 950XL. Many thanks

I still keep my Lumia 950XL. 
Could you provide me the instructions to install it in this device? Thank you.

**Comment by ch3mn3y** [comment]
@donho 
I have WOA installed. You cannot use it on Windows Mobile.
But WOA doesn't work that bad, have to say I work better than I expected, is quite usable and considering how many apps it support (both arm, as well as any x86 and x86_64 that don't need directX) its nice thing to play with if Lumia is not your mine device.
![IMG_20210405_075347](https://user-images.githubusercontent.com/20260224/113542232-1e65c400-95e4-11eb-9123-40a93abd7198.jpg)

**Comment by donho** [comment]
@ch3mn3y Very nice! Thank you!

**Comment by donho** [comment]
Hi guys, 
I need your help to test Notepad++ installer and for ARM64 (Signed as any official release package):
http://download.notepad-plus-plus.org/repository/MISC/ARM.test/npp.7.9.5.Installer.arm64.exe

Please let me know if the installer and binaries (notepad++.exe, plugins & auto-updater) work well under your ARM64 device. 

Thank you

**Comment by robmikh** [comment]
It appears that some of the plugins are still x86. I got these messages on launch for `mimeTools.dll`, `NppConverter.dll`, and `NppExport.dll`:

```
Cannot load 32-bit plugin.

<dll name> is not compatible with the current version of Notepad++.

Do you want to remove this plugin from the plugins directory to prevent this message from the next launch?
```

Additionally, the installer seems to want to put Notepad++ in the "Program Files (x86)" folder. Other than that Notepad++ itself was running and targeting ARM64.

**Comment by robmikh** [comment]
Actually, I'm not sure why that happened. I looked at the NT header for `NppConverter.dll` and it seems to say it's for ARM64...

![image](https://user-images.githubusercontent.com/7089228/115609776-d2ca4000-a29c-11eb-8926-bfef4a13743d.png)

[IMAGE_FILE_MACHINE_ARM64 ](https://docs.microsoft.com/en-us/windows/win32/sysinfo/image-file-machine-constants) is 0xAA64

**Comment by chcg** [comment]
@donho Seems at:

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/MISC/PluginsManager/PluginsManager.cpp#L30-L36

there needs to be an additional option for the ARM64 architecture, see https://docs.microsoft.com/en-us/windows/win32/debug/pe-format


`IMAGE_FILE_MACHINE_ARM64 | 0xaa64 | ARM64 little endian`

**Comment by donho** [comment]
@chcg Thank you a lot for your quick help. This commit https://github.com/notepad-plus-plus/notepad-plus-plus/commit/a4595236592c691b149326d56bc3fdb82dc57e6e should fix the not loading plugins bug.

@robmikh Thank you for your test and report. The new `npp.7.9.5.Installer.arm64.exe` is available in:
http://download.notepad-plus-plus.org/repository/MISC/ARM.test2/

Now the new installer should deploy the binaries in the right directory (for 64 bits programs), and the 3 plugins should be loaded and work.

**Comment by robmikh** [comment]
It worked flawlessly! The installer worked, and I did a simple test of the exporter plugin (export to HTML). There were no error messages on launch.

Thanks! 😄

**Comment by hmartinez82** [comment]
I can confirm it worked too, and it's loading JSON-Viewer from https://ci.appveyor.com/project/hernancmartinez/npp-json-viewer/build/job/203sd2v8is3k1ivb/artifacts . See (https://github.com/kapilratnani/JSON-Viewer/pull/106)

**Comment by donho** [comment]
Notepad++ v8 Release Candidate is available here:
https://community.notepad-plus-plus.org/topic/21228/notepad-v8-release-candidate-2

Could you guys try it out and confirm me it's working on your ARM64 device?
Thank you!

**Comment by savelov** [comment]
npp.8.0.Installer.arm64.exe works for me on ASUS TP370QL

**Comment by ntw-au** [comment]
Installer and app work for me also. My device is Samsung Galaxy Book2 running Windows 10 21H1. Thank you very much for your work on this port! 😊

**Comment by Qiangong2** [comment]
Any chance of getting a windows arm32 build?

**Comment by hmartinez82** [comment]
Tested and working on my Surface Pro X with Windows 21H1.

**Comment by hmartinez82** [comment]
It's only missing the Open with Notepad++ right-click context menu in Explorer

**Comment by AraHaan** [comment]
I wish Notepad++ could get a mac and linux version as well (since Mac supports X11).

**Comment by robmikh** [comment]
I'm late to the party but I grabbed the ARM64 release from the main site and it's working great.

**Comment by NiQ1** [comment]
> Notepad++ v8 Release Candidate is available here:
> https://community.notepad-plus-plus.org/topic/21228/notepad-v8-release-candidate-2
> 
> Could you guys try it out and confirm me it's working on your ARM64 device?
> Thank you!

By the way, a $75 Raspberry Pi 4 8GB is a great testing environment. Windows 10 ARM64 can be installed easily using the "Windows on Raspberry" installer. Not sure how legal it is but it will activate with retail keys.
